pkg/linter/lintRuleSyntacticRedundantCommon.go

```diff
@@ -0,0 +1,229 @@
+	for i, size := range tableToSize {
+		tableToRules[i] = make([]*vpcmodel.RuleOfFilter, size)
+	}
+	// 1.3 Populates tableToRules
+	for i := range rules {
+		filterIndex := rules[i].Filter.FilterIndex
+		tableToRules[filterIndex][rules[i].RuleIndex] = &rules[i]
+	}
+	// 2. For each table computes its atomic blocks and creates the above resulting map
+	tableToAtomicBlocks = map[int][]*ipblock.IPBlock{}
+	srcIPBlocks, dstIPBlocks := []*ipblock.IPBlock{}, []*ipblock.IPBlock{}
+	for tableIndex := range tableToRules {
+		for i := range tableToRules[tableIndex] {
+			srcIPBlocks = append(srcIPBlocks, tableToRules[tableIndex][i].SrcCidr)
+			dstIPBlocks = append(dstIPBlocks, tableToRules[tableIndex][i].DstCidr)
+		}
+		tableToAtomicBlocks[tableIndex] = ipblock.DisjointIPBlocks(srcIPBlocks, dstIPBlocks)
+	}
+	return tableToRules, tableToAtomicBlocks
+}
+
+/////////////////////////////////////////////////////////////
+//// finding interface implementation for ruleRedundant
+////////////////////////////////////////////////////////////
+
+func (finding *ruleRedundant) vpc() []vpcmodel.VPCResourceIntf {
+	return []vpcmodel.VPCResourceIntf{finding.vpcResource}
+}
+
+func (finding *ruleRedundant) string() string {
+	rule := finding.rule
+	strResPrefix := fmt.Sprintf("In VPC %s %s %s's rule %d is redundant. ",
+		finding.vpcResource.Name(), finding.rule.Filter.LayerName, rule.Filter.FilterName, rule.RuleIndex)
+	if rule.Filter.LayerName == NetworkACL {
+		if len(finding.containRules) == 1 {
+			strResPrefix += "It is shadowed by a higher priority rule"
+		} else { // >1
+			strResPrefix += "It is shadowed by higher priority rules"
+		}
+	} else {
+		if len(finding.containRules) == 1 {
+			strResPrefix += "It is implied by another rule"
+		} else { // >1
+			strResPrefix += "It is implied by other rules"
+		}
+	}
+	strResPrefix = strResPrefix + "\n\tRule's details: " + rule.RuleDesc
+	if rule.Filter.LayerName == NetworkACL {
+		strResPrefix += "\t\tShadowing rule"
+	} else {
+		strResPrefix += "\t\tImplying rule"
+	}
+	if len(finding.containRules) > 1 {
+		strResPrefix += "s:\n\t\t"
+	} else {
+		strResPrefix += ": "
+	}
+	containingRulesSlice := []string{}
+	for _, rule := range finding.containRules {
+		containingRulesSlice = append(containingRulesSlice, rule.RuleDesc)
+	}
+	sort.Strings(containingRulesSlice)
+	return strResPrefix + strings.Join(containingRulesSlice, "\t\t")
+}
+
+// for json:
+type ruleRedundantJSON struct {
+	Rule         vpcmodel.RuleOfFilter   `json:"vpc_name"`
+	VpcName      string                  `json:"rule_details"`
+	ContainRules []vpcmodel.RuleOfFilter // rules because of which this rule is redundant to their description
+}
+
+func (finding *ruleRedundant) toJSON() any {
+	rule := finding.rule
+	table := vpcmodel.Filter{LayerName: rule.Filter.LayerName,
+		FilterName: rule.Filter.FilterName}
+	containRules := make([]vpcmodel.RuleOfFilter, len(finding.containRules))
+	for i, rule := range finding.containRules {
```

Comment:
I think the i can be bigger than the size of containRules here

Reply:
Of course. You are much bigger ;-)

Reply:
done